Sources/Conduit/Networking/URLSessionClient.swift

```diff
@@ -315,18 +313,16 @@ private class SessionDelegate: NSObject, URLSessionDataDelegate {
         var responseData = taskResponse.data ?? Data()
         responseData.append(data)
         taskResponse.data = responseData
-        if let expectedContentLength = taskResponse.expectedContentLength,
-            let progressHandler = taskDownloadProgressHandlers[dataTask.taskIdentifier] {
-            let currentProgress = taskDownloadProgresses[dataTask.taskIdentifier]
-            let newProgress = currentProgress ?? Progress()
-            if currentProgress == nil {
-                serialQueue.sync {
-                    taskDownloadProgresses[dataTask.taskIdentifier] = newProgress
-                }
+        serialQueue.sync {
+            guard let expectedContentLength = taskResponse.expectedContentLength,
+                let progressHandler = taskDownloadProgressHandlers[dataTask.taskIdentifier] else {
+                    return
             }
-            newProgress.completedUnitCount = Int64(responseData.count)
-            newProgress.totalUnitCount = expectedContentLength
-            progressHandler(newProgress)
+            let downloadProgress = taskDownloadProgresses[dataTask.taskIdentifier] ?? Progress()
+            downloadProgress.completedUnitCount = Int64(responseData.count)
+            downloadProgress.totalUnitCount = expectedContentLength
+            taskDownloadProgresses[dataTask.taskIdentifier] = downloadProgress
+            progressHandler(downloadProgress)
```

Comment:
Should we throw `progressHandler` back outside of the internal queue? If `delegateQueue` has different attributes, or if it's otherwise important to the caller, then dispatching updates from a different queue would be unexpected

Reply:
It shouldn't be possible for two concurrent operations to be working on the same data task, right? Isn't `dataTask.identifier` unique to each request?

The problem we are solving on this PR is accessing the dictionaries, which are shared across requests in the same session. But tasks should not (or cannot) be shared.

Reply:
They aren't shared, but `SessionTaskProxy.downloadProgressHandler` and `SessionTaskProxy.uploadProgressHandler` are settable from client code. When a `URLSessionClient` is created with a `delegateQueue`, one might expect those handlers would be fired on that queue (which is where these delegate methods are called from)

Reply:
Oh, I see, will update, thanks!